### ISSUE TITLE
switch to poetry

### ISSUE THREAD

**Issue by skarzi** [seddonym/import-linter#80]
[`poetry`](https://python-poetry.org/) is becoming new standard when it comes to dependency management of python packages and it could help this great project in following ways:

+ one source of truth for all dependencies (also dev) - currently dev dependencies are defined in `tox.ini` and main in `setup.py`
+ easy way to run `black` or `isort` locally - when I was working on PR I had some troubles with applying `black` on the code - I have installed the freshest version from PyPI, but it formats some things diffrently than version used in the `tox - e check` that is listed is `tox.ini`. When this issue will be rejected, at least some more info should be added to `CONTRIBUTING.rst` 
+ easy way to add new dev dependencies like for instance `isort` or other that this project can benefit from 
+ easy integration with CI
+ easy integration with [`dependabot`](https://docs.github.com/en/github/administering-a-repository/keeping-your-dependencies-updated-automatically) and similar solutions, so always the most recent dependencies can be used (at least dev ones)
+ all features listed on `poetry` web page like easier publishing

**Comment by seddonym** [comment]
Thanks for this. For me, it sounds like the big win is that it might help others contribute to the project. 

I'd be happy to consider a PR if you felt it was worthwhile!

**Comment by skarzi** [comment]
Sure, I will try to prepare PR in following week! 

BTW why travis builds aren't visible as GitHub checks?  For instance in #79 there is no info about [travis builds](https://travis-ci.org/github/seddonym/import-linter/pull_requests)

**Comment by mwgamble** [comment]
I recommend using Renovatebot rather than Dependabot. It can group updates to avoid unnecessary PR noise. It's also more powerful and flexible, with the ability to maintain more dependencies on a schedule that suits you.

**Comment by bibz** [comment]
Since this issue was not solved and the WiP PR not completed, I'd like to chime in and advise _against_ Poetry and any other dependency management that caps dependency versions.
I agree they are nice tools that improve developer experience, but the fact that they, by default, restrict the upper version bound is ridiculous in the Python ecosystem. (See below for rationale.)

Maybe a set of recipes (Makefile) or tasks (e.g. [doit](https://pydoit.org/)) would improve the status quo while keeping the toolchain simple and based on well-known tools that do single things well?

---

Where I work, we have been bitten by Poetry a few times now. I started migrating away in favour of simpler tools such as _pip-tools_ (mainly because our libraries and apps are not published, so no need to upload them to PyPI otherwise [there are plenty other tools](https://packaging.python.org/en/latest/key_projects/)). Both for libraries and apps.

As a side effect, we have way fewer build dependencies (Poetry has _a lot_) which also speeds up our CI by a few minutes overall, I agree this is an implementation detail.

The missing pieces to the OP take the form of [PEP 517 – A build-system independent format for source trees](https://peps.python.org/pep-0517/) and [PEP 621 – Storing project metadata in pyproject.toml](https://peps.python.org/pep-0621/). Poetry's documentation does not even acknowledge the entries in `pyproject.toml` have been standardised.

---

I will let better writers (and known figures in the Python world) make the case:

- [Should You Use Upper Bound Version Constraints?](https://iscinumpy.dev/post/bound-version-constraints/) (and its [TL;DR](https://iscinumpy.dev/post/bound-version-constraints/#tldr)), _Henry Schreiner_
- [Semantic Versioning Will Not Save You](https://hynek.me/articles/semver-will-not-save-you/), _Hynek Schlawack_
- (YouTube) [_Antony Sottile_'s rant on Poetry](https://www.youtube.com/watch?v=Gr9o8MW_pb0)

**Comment by mwgamble** [comment]
> advise against Poetry and any other dependency management that caps dependency versions

Poetry doesn't force you to use the `^` operator for version pinning. It still accepts all of the same operators supported by setuptools, such as `>=`.

I really don't understand why this point is so offensive to people. Major versions carry the expectation of breaking changes. How can you possibly be certain your code is compatible with versions of packages that don't exist yet?

Pinning with `^` is the norm in PHP and Javascript. Why is Python special here?

> started migrating away in favour of simpler tools

"simple" is in the eye of the beholder. From my experience, "simple" usually means less powerful and fewer problems solved centrally. It usually means all the necessary complexity still exists, but is just pushed out elsewhere for other people to deal with.

> such as pip-tools

Where I work, we've been migrating away from pip-tools towards poetry:

- It's less powerful.
- Its dependency resolution algorithm isn't as good.
- It depends on undocumented internals of pip, which means it regularly breaks.
- Its approach to managing dev dependencies is subpar.
- Renovatebot (which is a significantly superior tool to Dependabot for automatically updating dependencies) has very poor support for pip-tools (if you can even call it support), but it has fantastic support for Poetry.
- It requires us to use a myriad of other tools for building distribution artifacts and publishing them to a central repository, whereas Poetry has built-in `build` and `publish` commands. This means it's significantly easier to teach people how to use Poetry (especially if they've come from a Javascript background), and significantly easier to manage scripts that invoke Poetry.

For an example of pip-tools' poor dependency resolution algorithm, see this issue, which is only affecting pip-tools: https://github.com/spulec/moto/issues/5485

> As a side effect, we have way fewer build dependencies (Poetry has a lot) which also speeds up our CI by a few minutes overall, I agree this is an implementation detail.

This can be solved by better caching of dependency installation. Especially for public projects using Github Actions, it's a relatively straightforward problem to solve.

> Poetry's documentation does not even acknowledge the entries in pyproject.toml have been standardised.

This is because Poetry existed before there was a standard for this. My personal opinion on the standard you've referenced is that it's an extremely poor kludge to just keep doing things the old way. It looks like nothing more than an incredibly thin wrapper around how setuptools and pip work right now. As someone who's come from other ecosystems where packaging isn't a total mess, I wish they'd throw out the old way rather than trying to keep build more cruft on top of it. Poetry represents the strive for a genuinely better future for Python package management, rather than just trying to paper over old problems with new problems.

Also, I think it's disingenuous to not acknowledge that the PEPs you referenced are being actively discussed in the Poetry issue tracker:

https://github.com/python-poetry/poetry/issues/3332

The documentation of a tool should reflect the current state of things, not a possible future that may never exist.

**Comment by bibz** [comment]
Thank you @mwgamble for your detailed answer, I appreciate some of your points but do not agree with everything.

In case it was not clear: I am not saying Poetry is bad, just that I would, personally, advise against using it because it promotes fragile behaviour by default. I acknowledge there is subjectivity here, I want to balance out the argument of "it's the new default standard" because it is *only* one tool, not a standard in any way, though it certainly did a lot of good to the Python packaging ecosystem by fostering different approaches and better DX.

TL;DR: I used my experience as an example, I am not saying it's the right and only way to go. I'd like to discuss the alternatives to Poetry.

---

> > advise against Poetry and any other dependency management that caps dependency versions
> 
> Poetry doesn't force you to use the `^` operator for version pinning. It still accepts all of the same operators supported by setuptools, such as `>=`.

Yes it does support the full spectrum of [PEP 508 – Dependency specification for Python Software Packages](https://peps.python.org/pep-0508/) and one can manually replace the caret versioning with a standard version spec. But let's be real: the whole point of a tool like Poetry is to automate this behaviour so that we do not manually tweak requirements by hand.

If I do `poetry add requests`, I don't want to then edit `pyproject.toml` to remove the caret spec. I'll trust Poetry to do the right thing (which is where I disagree with ^).

> I really don't understand why this point is so offensive to people. Major versions carry the expectation of breaking changes. How can you possibly be certain your code is compatible with versions of packages that don't exist yet?

I agree, you cannot. But that does not mean you **know** it will break either.
One of the points made in the (long) articles I listed previously is about the fact that SemVer is **not** a silver bullet, it is about communicating the developers intent ("expect breaking changes").
Obviously you need to run the test suite when updating, and maybe reviewing the changelogs too. But that's what you should do for minor and patch updates too.

According to SemVer, if a project is past `1.0.0`, you can expect some stability.
Have you witnessed complete API breakage when projects bump major versions? I have rarely, if at all, ever suffered that in 20 years of development.
Sure, it can happen, but it does not happen _often_ because 1) we usually only use a subset of a project API and 2) bumping major version does not mean everything breaks, just that _something_ changed (_usually_ one or two signatures).

> Pinning with `^` is the norm in PHP and Javascript. Why is Python special here?

I think pinning would be `==`, `^` is capping.
Pinning dependencies is certainly a good idea for reproducible-ish builds, and is a must-have for an app because you want to run with what you tested.

Again, referring to the articles, the difference between JS and Python is that dependencies are global in Python:

> The technical difference is that npm (and JavaScript) has the idea of local dependencies. If you have several packages that request the same dependency, they each get a copy of that dependency. They are free to have conflicting version requirements; each gets a copy so the copies could be different versions. 

In Python, each dependency you install **must** satisfy **all** requirements, including transitive requirements.
In Javascript, each dependency has its own cache of requirements, thus you can have different version of the same library installed in different places.

> > started migrating away in favour of simpler tools
> 
> "simple" is in the eye of the beholder. From my experience, "simple" usually means less powerful and fewer problems solved centrally. It usually means all the necessary complexity still exists, but is just pushed out elsewhere for other people to deal with.

Fair point.

> > such as pip-tools
> 
> Where I work, we've been migrating away from pip-tools towards poetry:
> 
>     * It's less powerful.
> 
>     * Its dependency resolution algorithm isn't as good.
> 
>     * It depends on undocumented internals of pip, which means it regularly breaks.
> 
>     * Its approach to managing dev dependencies is subpar.
> 
>     * Renovatebot (which is a significantly superior tool to Dependabot for automatically updating dependencies) has very poor support for pip-tools (if you can even call it support), but it has fantastic support for Poetry.
> 
>     * It requires us to use a myriad of other tools for building distribution artifacts and publishing them to a central repository, whereas Poetry has built-in `build` and `publish` commands. This means it's significantly easier to teach people how to use Poetry (especially if they've come from a Javascript background), and significantly easier to manage scripts that invoke Poetry.

I think this is where the discussion has a tint of subjectivity.
I _like_ the fact it's less powerful because it follows the KISS principle, I prefer my tools to be simpler (less complex) and do fewer things. I prefer composition. It seems you prefer a bigger tool that aggregates the behaviours. Nothing wrong with different tastes.

> Its dependency resolution algorithm isn't as good.

This is where it actually is better to *not* try to be smarter in my experience, and where Poetry failed us a few times: see [Backsolving is usually wrong](https://iscinumpy.dev/post/bound-version-constraints/#backsolving-is-usually-wrong).

> For an example of pip-tools' poor dependency resolution algorithm, see this issue, which is only affecting pip-tools: [spulec/moto#5485](https://github.com/spulec/moto/issues/5485)

I won't argue that `pip-tools` is one example, that it has issues, and that there are alternatives. It works for us and that's why I used it as _an example_. We have not encountered problems with it _so far_.

> > As a side effect, we have way fewer build dependencies (Poetry has a lot) which also speeds up our CI by a few minutes overall, I agree this is an implementation detail.
> 
> This can be solved by better caching of dependency installation. Especially for public projects using Github Actions, it's a relatively straightforward problem to solve.

Sure. Again, this is one example detail.
We cannot simply fix it that easily because Azure DevOps is full of papercuts (and arguably our projects have a weird setup), so it made a big difference to us.

> > Poetry's documentation does not even acknowledge the entries in pyproject.toml have been standardised.
> 
> This is because Poetry existed before there was a standard for this. My personal opinion on the standard you've referenced is that it's an extremely poor kludge to just keep doing things the old way. It looks like nothing more than an incredibly thin wrapper around how setuptools and pip work right now. As someone who's come from other ecosystems where packaging isn't a total mess, I wish they'd throw out the old way rather than trying to keep build more cruft on top of it. Poetry represents the strive for a genuinely better future for Python package management, rather than just trying to paper over old problems with new problems.

Oh yes, Python's packaging is a mess.
The general trend is to go slow and avoid Big Breaking Changes™ (*cough* Python 3 *cough*). It becomes a necessary evil to go slow and thoroughly discuss things.
Tools like Poetry can move faster and explore different avenues, and that's fantastic. It is also how standardisation happens.

I apologise, you're right, this was not a constructive criticism of Poetry's documentation.

> Also, I think it's disingenuous to not acknowledge that the PEPs you referenced are being actively discussed in the Poetry issue tracker:
> 
> [python-poetry/poetry#3332](https://github.com/python-poetry/poetry/issues/3332)
> 
> The documentation of a tool should reflect the current state of things, not a possible future that may never exist.

No hard bashing intended, Poetry is actively developed and I missed that discussion.

**Comment by mwgamble** [comment]
> because it promotes fragile behaviour by default

I have to argue quite the opposite. Promoting the use of unconstrained dependencies that allow automatic upgrades to broken versions (ie. new major versions) seems significantly more fragile to me.

**Comment by mwgamble** [comment]
> This is where it actually is better to not try to be smarter in my experience, and where Poetry failed us a few times: see [Backsolving is usually wrong](https://iscinumpy.dev/post/bound-version-constraints/#backsolving-is-usually-wrong).

The whole premise of that section of that article is based on the fact that your dependencies are themselves using unconstrained dependencies. This just proves it's an ecosystem-wide problem. If everyone was actually properly constraining their dependencies properly (like what happens in other language ecosystems), this wouldn't be a concern, and backsolving would absolutely be a great solution.

There are other things from other packaging ecosystems that are desperately missing for Python. For example, there's no way to say "my package conflicts with this other package" or "my package conflicts with these versions of this other package". These would also help solve these problems.

**Comment by seddonym** [comment]
Thanks for the discussion!

I have to admit I've also come across [Antony Sottile's rant on Poetry](https://www.youtube.com/watch?v=Gr9o8MW_pb0) since my original comment, which makes me less sure this is the right direction for Import Linter.

The truth is that dependency management is not much of a pain point for Import Linter at present. I appreciate using Poetry might make contributing a bit easier for some people, though it might make it harder for others who are unfamiliar with the tool. So I think it's unlikely - at least at the moment - that this will happen. Never say never though!

**Comment by seddonym** [comment]
I'm going to close this ticket for now as I don't currently intend to move to Poetry. Thanks everyone for your contributions!
